internal/mesh/internal/controllers/sidecar-proxy/controller.go

```diff
@@ -0,0 +1,168 @@
+	"github.com/hashicorp/consul/internal/mesh/internal/controllers/sidecar-proxy/cache"
+	"github.com/hashicorp/consul/internal/mesh/internal/controllers/sidecar-proxy/fetcher"
+	"github.com/hashicorp/consul/internal/mesh/internal/controllers/sidecar-proxy/mapper"
+	"github.com/hashicorp/consul/internal/mesh/internal/types"
+	"github.com/hashicorp/consul/internal/mesh/internal/types/intermediate"
+	"github.com/hashicorp/consul/internal/resource"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
+)
+
+// ControllerName is the name for this controller. It's used for logging or status keys.
+const ControllerName = "consul.io/sidecar-proxy-controller"
+
+type TrustDomainFetcher func() (string, error)
+
+func Controller(cache *cache.Cache, mapper *mapper.Mapper, trustDomainFetcher TrustDomainFetcher) controller.Controller {
+	if cache == nil || mapper == nil || trustDomainFetcher == nil {
+		panic("cache, mapper and trust domain fetcher are required")
+	}
+
+	return controller.ForType(types.ProxyStateTemplateType).
+		WithWatch(catalog.ServiceEndpointsType, mapper.MapServiceEndpointsToProxyStateTemplate).
+		WithWatch(types.UpstreamsType, mapper.MapDestinationsToProxyStateTemplate).
+		WithReconciler(&reconciler{cache: cache, getTrustDomain: trustDomainFetcher})
+}
+
+type reconciler struct {
+	cache          *cache.Cache
+	getTrustDomain TrustDomainFetcher
+}
+
+func (r *reconciler) Reconcile(ctx context.Context, rt controller.Runtime, req controller.Request) error {
+	rt.Logger = rt.Logger.With("resource-id", req.ID, "controller", ControllerName)
+
+	rt.Logger.Trace("reconciling proxy state template", "id", req.ID)
+
+	// Instantiate a data fetcher to fetch all reconciliation data.
+	dataFetcher := fetcher.Fetcher{Client: rt.Client, Cache: r.cache}
+
+	// Check if the workload exists.
+	workloadID := resource.ReplaceType(catalog.WorkloadType, req.ID)
+	workload, err := dataFetcher.FetchWorkload(ctx, resource.ReplaceType(catalog.WorkloadType, req.ID))
+	if err != nil {
+		rt.Logger.Error("error reading the associated workload", "error", err)
+		return err
+	}
+	if workload == nil {
+		// If workload has been deleted, then return as ProxyStateTemplate should be cleaned up
+		// by the garbage collector because of the owner reference.
+		rt.Logger.Trace("workload doesn't exist; skipping reconciliation", "workload", workloadID)
+		return nil
+	}
+
+	proxyStateTemplate, err := dataFetcher.FetchProxyStateTemplate(ctx, req.ID)
+	if err != nil {
+		rt.Logger.Error("error reading proxy state template", "error", err)
+		return nil
+	}
+
+	if proxyStateTemplate == nil {
+		// If proxy state template has been deleted, we will need to generate a new one.
+		rt.Logger.Trace("proxy state template for this workload doesn't yet exist; generating a new one", "id", req.ID)
+	}
+
+	if !fetcher.IsMeshEnabled(workload.Workload.Ports) {
```

Comment:
I suspect you will want to delete the old resource should the linked workload lose it's mesh port, as the owner ref will not handle that for you.